### PR TITLE
Move type definitions of pubkey/script hash types

### PR DESCRIPTION
**Pull request by tcharding**
Total re-write since review. Now this PR moves the hash type definitions out of `hash_types`. Please see https://github.com/rust-bitcoin/rust-bitcoin/issues/1909#issuecomment-1603634440 for more.

No longer adds unit tests.

Fix: #1909

### CODE REVIEW

**Review by apoelstra — DISMISSED** [review]
ACK ee33eefe04ef9aa809904b4c42c41f9cf2caf647

**Review by Kixunil — COMMENTED** [review]
Wouldn't this be better as doctest?

**Comment by yancyribbens** [comment]
Also From is defined in [key.rs](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/crypto/key.rs#L287) so the test could go there?  I don't really understand why the impl PubkeyHash and the struct are in different files along with the traits.

**Comment by tcharding** [comment]
> I don't really understand why the impl PubkeyHash and the struct are in different files along with the traits.

BOOM! Thanks man.

File: bitcoin/src/crypto/key.rs [hunk 1]
```diff
@@ -116,8 +116,7 @@ impl PublicKey {
                 hash160::Hash::hash(&self.inner.serialize()).to_byte_array(),
             ))
         } else {
-            // We can't create witness pubkey hashes for an uncompressed
-            // public keys
+            // We can't create witness pubkey hash for an uncompressed pubkey.
```
Comment (apoelstra) [inline bitcoin/src/crypto/key.rs:118]: In 76f9ab2df76d5a65ade52c339d51a2999b8446ab

Should be "a witness pubkey hash".

**Comment by apoelstra** [comment]
You say `new` does not make sense for `Script` since it implies ownership, but both `Path` and `PathBuf` have `new` constructors (and neither have `empty` one).

I think we should use the name `new` for both.

**Comment by apoelstra** [comment]
ack b798939fde1d9890a93d41abe31eaf23e4a5356e other than the `new`/`empty` thing. Also, I'm unconvinced that moving the hash types is a breaking change, but it's fine, since the `Script` constructor rename definitely will be.

**Comment by tcharding** [comment]
Will fix  emty/new thing as suggested.

> I'm unconvinced that moving the hash types is a breaking change

Well after this one can no longer write `use bitcoin::hash_types::PubkeyHash` and before this they can. That is a breaking change, right? IIRC I mentioned it because I intentionally did not add a re-export in `hash_types` since we are moving loads of things out of there.

File: bitcoin/src/crypto/key.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/crypto/key.rs:118]: Done, thanks.

**Comment by tcharding** [comment]
I just realized the new/empty change should not even be in this PR, its totally unrelated to pubkey/script hash - will pull it out into a separate PR.

**Comment by tcharding** [comment]
Added a patch `9216ab11 Move re-export of newtypes::*` to stop rustfmt messing with the code layout.

**Comment by tcharding** [comment]
Rebased on master (to pick up #1927). No other changes.

**Review by apoelstra — DISMISSED** [review]
ACK dcfbafd75572801e1fd8582c165a3cf956b0d80b

**Comment by tcharding** [comment]
Removed the first two patches (because we now have https://github.com/rust-bitcoin/rust-bitcoin/pull/1925) and rebased.

**Comment by tcharding** [comment]
Pulled master and rebased, woops.

**Review by apoelstra — APPROVED** [review]
ACK 10a95a6cf98c97f397affedfb44f1bdaca1b2178

**Review by apoelstra — APPROVED** [review]
ACK 27b3c1e0e6bc192412e1c625d913b3e2651167e6

**Comment by apoelstra** [comment]
Merging based on #1945. The rebase made no changes.

**Pull request merged**